README.md
[![Build Status](https://travis-ci.org/nnutter/mojo-pgx-cursor.svg?branch=master)](https://travis-ci.org/nnutter/mojo-pgx-cursor)
# NAME

Mojo::PgX::Cursor - Cursor Extension for Mojo::Pg

# SYNOPSIS

    require Mojo::PgX::Cursor;
    my $pg = Mojo::PgX::Cursor->new('postgresql://postgres@/test');
    my $results = $pg->db->cursor('select * from some_big_table');
    while (my $next = $results->hash) {
      say $next->{name};
    }

# DESCRIPTION

[DBD::Pg](https://metacpan.org/pod/DBD::Pg) fetches all rows when a statement is executed whereas other drivers
usually fetch rows using the `fetch*` methods.  `Mojo::PgX::Cursor` is an
extension to work around this issue using PostgreSQL cursors while providing a
[Mojo::Pg](https://metacpan.org/pod/Mojo::Pg)-style API for iteratoring over the results; see
[Mojo::PgX::Cursor::Results](https://metacpan.org/pod/Mojo::PgX::Cursor::Results) for details.

# METHODS

## db

This subclass overrides [Mojo::Pg](https://metacpan.org/pod/Mojo::Pg)'s implementation in order to subclass the
resulting [Mojo::Pg::Database](https://metacpan.org/pod/Mojo::Pg::Database) object into a [Mojo::PgX::Cursor::Database](https://metacpan.org/pod/Mojo::PgX::Cursor::Database).

# VERSIONING

This module will follow [Semantic Versioning
2.0.0](http://semver.org/spec/v2.0.0.html).  Once the API feels reasonable I'll
release v1.0.0 which would correspond to 1.000000 according to [version](https://metacpan.org/pod/version),

    version->declare(q(v1.0.0))->numify # 1.000000
    version->parse(q(1.000000))->normal # v1.0.0

# MONKEYPATCH

    require Mojo::Pg;
    require Mojo::PgX::Cursor;
    use Mojo::Util 'monkey_patch';
    monkey_patch 'Mojo::Pg::Database', 'cursor', \&Mojo::PgX::Cursor::Database::cursor;

Just because you can doesn't mean you should but if you want you can
`monkey_patch` [Mojo::Pg::Database](https://metacpan.org/pod/Mojo::Pg::Database) rather than swapping out your
construction of [Mojo::Pg](https://metacpan.org/pod/Mojo::Pg) objects with the [Mojo::PgX::Cursor](https://metacpan.org/pod/Mojo::PgX::Cursor) subclass.

# DISCUSSION

This module would be unnecessary if [DBD::Pg](https://metacpan.org/pod/DBD::Pg) did not fetch all rows during
`execute` and since `libpq` supports that it would be much better to fix
`fetch*` than to implement this.  However, I am not able to do so at this
time.

# CONTRIBUTING

If you would like to submit bug reports, feature requests, questions, etc. you
should create an issue on the [GitHub Issue
Tracker](https://github.com/nnutter/mojo-pgx-cursor/issues) for this module.

# REFERENCES

- [#93266 for DBD-Pg: DBD::Pg to set the fetch size](https://rt.cpan.org/Public/Bug/Display.html?id=93266)
- [#19488 for DBD-Pg: Support of cursor concept](https://rt.cpan.org/Public/Bug/Display.html?id=19488)

# LICENSE

Copyright (C) Nathaniel Nutter.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Nathaniel Nutter `nnutter@cpan.org`

# SEE ALSO

[DBD::Pg](https://metacpan.org/pod/DBD::Pg), [Mojo::Pg](https://metacpan.org/pod/Mojo::Pg), [Mojo::PgX::Cursor::Cursor](https://metacpan.org/pod/Mojo::PgX::Cursor::Cursor),
[Mojo::PgX::Cursor::Database](https://metacpan.org/pod/Mojo::PgX::Cursor::Database), [Mojo::PgX::Cursor::Results](https://metacpan.org/pod/Mojo::PgX::Cursor::Results)


Changes
Revision history for Perl extension Mojo-PgX-Cursor

{{$NEXT}}

0.502001 2016-02-06T18:52:14Z

    - Fix "circular reference" that prevented proper destruction of
      Mojo::PgX::Cursor cursors.  See `t/scope-destroy-example.t`.

0.502000 2015-10-26T02:44:54Z

    - Add `wait` method on Mojo::PgX::Cursor::Results.
    - Remove `fetch_count` attribute on Mojo::PgX::Cursor::Results in favor of
      overloading `rows` to set the fetch count.

0.501000 2015-10-25T03:27:57Z

    - Rename `fetch` attribute on Mojo::PgX::Cursor::Results to `fetch_count`.
    - Documentation improvements.

0.500001 2015-10-24T21:49:36Z

    - Fix race condition when trying to access `rows` or `columns` before
      results have been fetched.

0.500000 2015-10-24T07:08:26Z

    - Add documentation about versioning.
    - Suppress PostgreSQL notices in tests.

0.04 2015-10-24T06:42:44Z

    - Specify implicit dependencies.
    - Specify required versions of dependencies.

0.03 2015-10-24T05:28:43Z

    - Manually created cursors now also default to fetching 100 row iterations.
    - Fix warning during cursor destruction.
    - Documentation cleanup.
    - Internal code cleanup.

0.02 2015-10-24T03:56:38Z

    - Results pre-fetch next batch immediately rather than waiting for first
      row of current batch to be pulled off.
    - Increase default fetch size to 100.
    - Remove the 'reload_at' attribute; it was unnecessary complexity.
    - Remove the 'monkey_patch' import option.  If you wish to monkey patch
      Mojo::Pg then see the documentation on how to do so.

0.01 2015-10-23T02:15:50Z

    - original version


examples/perf.pl
use Mojo::Base -strict;

use Mojo::JSON qw(encode_json);

unless ($ENV{TEST_ONLINE}) {
  say STDERR 'set TEST_ONLINE to run this';
  exit 1;
}

require Mojo::PgX::Cursor;;

my $pg = Mojo::PgX::Cursor->new($ENV{TEST_ONLINE});
my $db = $pg->db;
my $results = eval { $db->query('select count(*) from perf_test') };
unless ($results) {
  $db->query(
    'create table if not exists perf_test (
    id   serial primary key,
    name text,
    jdoc jsonb
    )'
  );

  my $tx = $db->begin;
  my @rows = (
    ['foo', encode_json { value => $_ }],
    ['bar', encode_json { value => $_ }],
  );
  for (1..500_000) {
    for (@rows) {
      $tx->db->query('insert into perf_test (name, jdoc) values (?, ?)', $_->[0], $_->[1]);
    }
  }
  $tx->commit;
}

use Time::HiRes qw(nanosleep time);
for my $rows (reverse (1, 10, 100, 1000, 10000)) {
  my $start = time;
  my $cursor = $db->cursor('select * from perf_test limit 100000');
  $cursor->rows($rows);
  while (my $row = $cursor->hash) {
    nanosleep 10;
  }
  my $elapsed = time - $start;
  say sprintf 'Blocked for %6.3f seconds (%4.1f%%) with rows = %5d',
    $cursor->seconds_blocked, ($cursor->seconds_blocked / $elapsed * 100), $rows;
  $cursor->wait;
}


lib/Mojo/PgX/Cursor/Results.pm
package Mojo::PgX::Cursor::Results;

require Mojo::IOLoop;
require Mojo::PgX::Cursor::Cursor;

use Time::HiRes qw(time);

use Mojo::Base -base;

has [qw(seconds_blocked)];

sub array {
  my $self = shift->_fetch;
  $self->{remaining}--;
  return $self->{results}->array;
}

sub columns { shift->_results->columns }

sub cursor {
  my $self = shift;
  if (@_) {
    $self->wait;
    $self->{cursor} = shift;
    $self->{remaining} = 0;
    delete $self->{delay};
    delete $self->{next};
    delete $self->{results};
    return $self->_load_next;
  }
  return $self->{cursor};
}

sub expand {
  my $self = shift;
  $self->{expand}++;
  if ($self->{results}) {
      $self->{results}->expand;
  }
  return $self;
}

sub hash {
  my $self = shift->_fetch;
  $self->{remaining}--;
  return $self->{results}->hash;
}

sub new {
  my $self = shift->SUPER::new(@_);
  return $self->_load_next;
}

sub rows {
  my $self = shift;
  return $self->_results->rows unless @_;
  $self->{rows} = shift;
  return $self;
}

sub wait {
  my $self = shift;
  if ($self->{delay}) {
    $self->{delay}->wait;
  }
  return $self;
}

sub _fetch {
  my $self = shift;
  return $self if $self->{remaining};
  unless ($self->{next}) {
    my $start = time;
    $self->wait;
    $self->{seconds_blocked} += time - $start;
    delete $self->{delay};
  }
  $self->{results} = delete $self->{next};
  $self->{remaining} = $self->{results}->rows;
  return $self->_load_next;
}

sub _load_next {
  my $self = shift;
  $self->{delay} = Mojo::IOLoop->delay(
    sub {
      $self->cursor->fetch($self->{rows}, shift->begin);
    },
    sub {
      $self->{next} = $_[2];
      $self->{next}->expand if ($self->{expand});
    },
  );
  return $self;
}

sub _results {
  my $self = shift;
  return $self->{results} if $self->{results};
  return $self->_fetch->{results};
}

1;
__END__

=encoding utf-8

=head1 NAME

Mojo::PgX::Cursor::Results

=head1 DESCRIPTION

L<Mojo::PgX::Cursor::Results> is a container for a L<Mojo::PgX::Cursor::Cursor>
like L<Mojo::Pg::Results> is for a statement handle.  Therefore it tries to
mimic the API of L<Mojo::Pg::Results> whereever it makes sense to do do.

This container should behave like L<Mojo::Pg::Results> for C<array>,
C<columns>, C<hash>, and C<expand>.  It intentionally does not support
C<hashes> or C<arrays> since if you wish to use those you should just use
C<query> instead.  One difference in behavior is C<rows> returns the number of
rows in the current iteration not the total rows for the query.

=head1 ATTRIBUTES

=head2 cursor

    my $cursor = $results->cursor;
    $results = $results->cursor($cursor);

The L<Mojo::PgX::Cursor::Cursor> rows are fetched from.

=head2 seconds_blocked

    my $time_wasted = $results->seconds_blocked;

The cumulative time the cursor has spent waiting for rows.

=head1 METHODS

=head2 array

    my $row = $results->array;

Return next row from L</"cursor"> as an array reference.

=head2 columns

    my $columns = $results->columns;

Return column names as an array reference.

=head2 expand

    $results = $results->expand;

Decode C<json> and C<jsonb> fields automatically for all rows.

=head2 hash

    my $row = $results->hash;

Return next row from L</"cursor"> as a hash reference.

=head2 new

    my $results = Mojo::PgX::Cursor::Results->new(cursor => $cursor);

Construct a new L<Mojo::PgX::Cursor::Results> object.

=head2 rows

    my $num = $results->rows;

Number of rows in current iteration; not the total for the original query.

    $results->rows(1000);

Set the number of rows to fetch in each iteration.  Since the next iteration is
always pre-fetched up to twice this many rows will be in memory at any given
time.  Set this to optimize for time or memory in your specific use case.

If not set then the L<Mojo::PgX::Cursor::Cursor> default fetch size is used.

=head2 wait

  $results->wait;

Wait for background query to finish.  This would be used to block before making
another background query on the same database handle.

=head1 LICENSE

Copyright (C) Nathaniel Nutter.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Nathaniel Nutter C<nnutter@cpan.org>

=head1 SEE ALSO

L<Mojo::PgX::Cursor>, L<Mojo::PgX::Cursor::Cursor>, L<Mojo::PgX::Cursor::Database>

=cut



lib/Mojo/PgX/Cursor/Database.pm
package Mojo::PgX::Cursor::Database;

require Mojo::PgX::Cursor::Cursor;
require Mojo::PgX::Cursor::Results;

use Mojo::Base 'Mojo::Pg::Database';

sub cursor {
  my ($self, $query, @bind) = (shift, shift, @_);
  my $cursor = Mojo::PgX::Cursor::Cursor->new(
    query => $query,
    db    => $self,
    bind  => \@bind,
  );
  return Mojo::PgX::Cursor::Results->new(cursor => $cursor);
}

1;
__END__

=encoding utf-8

=head1 NAME

Mojo::PgX::Cursor::Database

=head1 DESCRIPTION

Subclass of L<Mojo::Pg::Database>.  Adds the C<cursor> method.

=head1 METHODS

=head2 cursor

    my $results = $db->cursor('select * from foo');
    my $results = $db->cursor('select * from foo where id >= (?)', 10);

Create a PostgreSQL cursor and return an L<Mojo::PgX::Cursor::Results> object.

=head1 LICENSE

Copyright (C) Nathaniel Nutter.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Nathaniel Nutter C<nnutter@cpan.org>

=head1 SEE ALSO

L<Mojo::PgX::Cursor>, L<Mojo::PgX::Cursor::Cursor>, L<Mojo::PgX::Cursor::Results>

=cut


lib/Mojo/PgX/Cursor.pm
package Mojo::PgX::Cursor;

require Mojo::PgX::Cursor::Database;

use Mojo::Base 'Mojo::Pg';

our $VERSION = "0.502000";

sub db {
    my $db = shift->SUPER::db(@_);
    return bless $db, 'Mojo::PgX::Cursor::Database';
}

1;
__END__

=encoding utf-8

=head1 NAME

Mojo::PgX::Cursor - Cursor Extension for Mojo::Pg

=head1 SYNOPSIS

    my $pg = Mojo::PgX::Cursor->new('postgresql://postgres@/test');
    my $results = $pg->db->cursor('select * from some_big_table');
    while (my $next = $results->hash) {
      say $next->{name};
    }

=head1 DESCRIPTION

L<DBD::Pg> fetches all rows when a statement is executed whereas other drivers
usually fetch rows using the C<fetch*> methods.  Mojo::PgX::Cursor is an
extension to work around this issue using PostgreSQL cursors while providing a
L<Mojo::Pg>-style API for iteratoring over the results; see
L<Mojo::PgX::Cursor::Results> for details.

=head1 METHODS

=head2 db

This subclass overrides L<Mojo::Pg>'s implementation in order to subclass the
resulting L<Mojo::Pg::Database> object into a L<Mojo::PgX::Cursor::Database>.

=head1 VERSIONING

This module will follow L<Semantic Versioning
2.0.0|http://semver.org/spec/v2.0.0.html>.  Once the API feels reasonable I'll
release v1.0.0 which would correspond to 1.000000 according to L<version>,

    version->declare(q(v1.0.0))->numify # 1.000000
    version->parse(q(1.000000))->normal # v1.0.0

=head1 MONKEYPATCH

    require Mojo::Pg;
    require Mojo::PgX::Cursor;
    use Mojo::Util 'monkey_patch';
    monkey_patch 'Mojo::Pg::Database', 'cursor', \&Mojo::PgX::Cursor::Database::cursor;

Just because you can doesn't mean you should but if you want you can
C<monkey_patch> L<Mojo::Pg::Database> rather than swapping out your
construction of L<Mojo::Pg> objects with the L<Mojo::PgX::Cursor> subclass.

=head1 DISCUSSION

This module would be unnecessary if L<DBD::Pg> did not fetch all rows during
C<execute> and since C<libpq> supports that it would be much better to fix
C<fetch*> than to implement this.  However, I am not able to do so at this
time.

=head1 CONTRIBUTING

If you would like to submit bug reports, feature requests, questions, etc. you
should create an issue on the L<GitHub Issue
Tracker|https://github.com/nnutter/mojo-pgx-cursor/issues> for this module.

=head1 REFERENCES

=over

=item L<#93266 for DBD-Pg: DBD::Pg to set the fetch size|https://rt.cpan.org/Public/Bug/Display.html?id=93266>

=item L<#19488 for DBD-Pg: Support of cursor concept|https://rt.cpan.org/Public/Bug/Display.html?id=19488>

=back

=head1 LICENSE

Copyright (C) Nathaniel Nutter.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Nathaniel Nutter C<nnutter@cpan.org>

=head1 SEE ALSO

L<DBD::Pg>, L<Mojo::Pg>, L<Mojo::PgX::Cursor::Cursor>,
L<Mojo::PgX::Cursor::Database>, L<Mojo::PgX::Cursor::Results>

=cut

